### PR TITLE
Class Unique 

### PR DESCRIPTION
**Pull request by harlouci**
I added Unique class

### CODE REVIEW

File: theano/tensor/extra_ops.py [hunk 1]
```diff
@@ -2,6 +2,7 @@
 import numpy
 import warnings
 import theano
+import theano.tensor as T
```
Comment (nouiz) [inline theano/tensor/extra_ops.py]: remove and use basic.* instead

File: theano/tensor/extra_ops.py [hunk 2]
```diff
@@ -1006,3 +1007,69 @@ def to_one_hot(y, nb_class, dtype=None):
     ret = theano.tensor.set_subtensor(ret[theano.tensor.arange(y.shape[0]), y],
                                       1)
     return ret
+
+class Unique(theano.Op):
+    __props__ = ("return_index", "return_inverse", "return_counts")
+
+    def __init__(self, return_index=False, return_inverse=False, 
+				 return_counts=False):
```
Comment (nouiz) [inline theano/tensor/extra_ops.py]: indentation. always use only space, not tab.

File: theano/tensor/extra_ops.py [hunk 3]
```diff
@@ -1006,3 +1007,69 @@ def to_one_hot(y, nb_class, dtype=None):
...
+        #x = x.flatten()
+        
+        outputs = []
+        # output0 = T.TensorType(broadcastable=[False], dtype=x.dtype)()
+        output0 = x.flatten().type()
+        outputs.append(output0)
+        typ = T.TensorType(broadcastable=[False], dtype='int64')
+        if self.return_index :
+            output1 = typ()
+            outputs.append(output1)
```
Comment (nouiz) [inline theano/tensor/extra_ops.py]: merge those 2 lines in 1 line. Idem bellow.

File: theano/tensor/extra_ops.py [hunk 4]
```diff
@@ -1006,3 +1007,69 @@ def to_one_hot(y, nb_class, dtype=None):
...
+        if self.return_inverse :
+            output2 = typ()#T.TensorType(broadcastable=[False], dtype=x.dtype)
+            outputs.append(output2)
+        if self.return_counts :
+            output3 = typ()#T.TensorType(broadcastable=[False], dtype=x.dtype)
```
Comment (nouiz) [inline theano/tensor/extra_ops.py]: remove the not useful comment. Idem above.

File: theano/tensor/extra_ops.py [hunk 5]
```diff
@@ -1006,3 +1007,69 @@ def to_one_hot(y, nb_class, dtype=None):
...
+		        (not self.return_counts)):
+            z[0][0]=outs
+        else :
+            for i in range(len(outs)): 
+                z[i][0] = outs[i]
+                
+    def infer_shape(self, node, i0_shapes):
+        ret = node.fgraph.shape_feature.default_infer_shape(node, i0_shapes)
+        if self.return_inverse :
+            shape = (np.prod(i0_shapes[0]), )
```
Comment (nouiz) [inline theano/tensor/extra_ops.py]: use basic.prod(), this is symbolic shape. np.prod() should work, but make it less clear we work on symbolic shape.

File: theano/tensor/tests/test_extra_ops.py [hunk 6]
```diff
@@ -2,8 +2,8 @@
 import numpy as np
 import numpy
 import unittest
-
```
Comment (nouiz) [inline theano/tensor/tests/test_extra_ops.py]: keep that empty line.

File: theano/tensor/tests/test_extra_ops.py [hunk 7]
```diff
@@ -2,8 +2,8 @@
...
 import theano
+
```
Comment (nouiz) [inline theano/tensor/tests/test_extra_ops.py]: remove this one.

File: theano/tensor/tests/test_extra_ops.py [hunk 8]
```diff
@@ -661,3 +661,66 @@ def test_to_one_hot():
...
          [0., 0., 0., 0., 0., 1., 0., 0., 0., 0.],
          [0., 0., 0., 0., 0., 0., 1., 0., 0., 0.]])
+
+class test_Unique(utt.InferShapeTester):
+    
+    def setUp(self):
+        super(test_Unique, self).setUp()
+        self.op_class = Unique
+        self.ops = [Unique(), Unique(True), Unique(True, True), 
+					Unique(False, True)]#, Unique(True, True, True)]
```
Comment (nouiz) [inline theano/tensor/tests/test_extra_ops.py]: Add the last case is the numpy version is recent enought.

File: theano/tensor/tests/test_extra_ops.py [hunk 9]
```diff
@@ -661,3 +661,66 @@ def test_to_one_hot():
...
+                              [[1., 2., 3.], [1, 0, 2], [1, 0, 2, 1]],
+                              [[1., 2., 3.], [1, 0, 2, 1]]]
+        for op, outs_expected in zip(self.ops, list_outs_expected) :
+            f = theano.function(inputs=[x], outputs=op(x, return_list=True))
+            outs = f(inp)
+            print outs
+            # Compare the result computed to the expected value.
+            for out, out_exp in zip(outs, outs_expected):
+                print out
+                print out_exp
```
Comment (nouiz) [inline theano/tensor/tests/test_extra_ops.py]: remove the prints, we don't want tests to be verbose by defaults.

**Comment by nouiz** [comment]
thanks for the PR, I did a first round of review.

**Comment by harlouci** [comment]
2eme pull

**Comment by carriepl** [comment]
Some tests are failing under Travis. Do you know how to check which ones?
Also, to facilitate review, please take the habit of giving meaningful names to your pull requests and your commits.

**Comment by harlouci** [comment]
*$ python extra_ops.py*
Traceback (most recent call last):
  File "extra_ops.py", line 4, in <module>
    import theano
  File "/u/harlouci/LISA/Theano/Theano/theano/__init__.py", line 44, in
<module>
    from theano.gof import (
  File "/u/harlouci/LISA/Theano/Theano/theano/gof/__init__.py", line 38, in
<module>
    from theano.gof.cc import \
  File "/u/harlouci/LISA/Theano/Theano/theano/gof/cc.py", line 59, in
<module>
    from theano.gof import link
  File "/u/harlouci/LISA/Theano/Theano/theano/gof/link.py", line 13, in
<module>
    from theano.gof.type import Type
  File "/u/harlouci/LISA/Theano/Theano/theano/gof/type.py", line 11, in
<module>
    from theano.gof.op import CLinkerObject
  File "/u/harlouci/LISA/Theano/Theano/theano/gof/op.py", line 31, in
<module>
    from theano.gof.cmodule import GCC_compiler
  File "/u/harlouci/LISA/Theano/Theano/theano/gof/cmodule.py", line 13, in
<module>
    import tempfile
  File
"/u/lisa/local/install_package.fc14_v3/Canopy1.4/appdata/canopy-1.4.0.1938.rh5-x86_64/lib/python2.7/tempfile.py",
line 32, in <module>
    import io as _io
  File "/u/harlouci/GitHub/Theano/theano/tensor/io.py", line 2, in <module>
    from theano import gof
ImportError: cannot import name gof

[harlouci@eos3 *theano*]$ ls
compat              configparser.py   gradient.pyc  misc
sandbox      tensor      updates.pyc
compile             configparser.pyc  ifelse.py     printing.py
scalar       tests       version.py
configdefaults.py   *gof*               __init__.py   printing.pyc
scan_module  typed_list  version.pyc
configdefaults.pyc  gradient.py       __init__.pyc  raise_op.py
sparse       updates.py



2015-06-09 14:47 GMT-04:00 carriepl <notifications@github.com>:

**Comment by harlouci** [comment]
Bon, je laisse tomber provisoirement, car j'obtiens le même message
d'erreur.
Merci tout de même,
Iban

2015-06-10 14:43 GMT-04:00 Harlouchet Iban <iban.harlouchet@gmail.com>:

**Comment by dwf** [comment]
Just a general comment: it is preferable to use pull request names, pull request descriptions and commit messages that are descriptive of the work being done. C'est pas grave pour moi si les messages sont en français ou en anglais (les collaborateurs principaux utilisent souvent l'anglais sur GitHub, peut-être ils ont une position différente) mais idéalement quelqu'un devrait être capable de comprendre gênéralement le sujet du PR ou des commits individuels par les messages correspondants, même si un lecteur pas francophone doit utiliser Google Translate.

**Comment by harlouci** [comment]
Correction of infer_shape tests

File: theano/tensor/extra_ops.py [hunk 10]
```diff
@@ -1006,3 +1006,65 @@ def to_one_hot(y, nb_class, dtype=None):
...
+
+class Unique(theano.Op):
+    __props__ = ("return_index", "return_inverse", "return_counts")
+
+    def __init__(self, return_index=False, return_inverse=False, 
+                 return_counts=False):
+        self.return_index = return_index
+        self.return_inverse = return_inverse
+        self.return_counts = return_counts
+        if self.return_counts == True and np.__version__ < "1.9.0" :
```
Comment (nouiz) [inline theano/tensor/extra_ops.py]: The right way to compare the version number is:

        numpy_ver = [int(n) for n in numpy.__version__.split('.')[:2]]
        if bool(numpy_ver < [1, 9]):
Do the change for all check.

File: theano/tensor/extra_ops.py [hunk 11]
```diff
@@ -1006,3 +1006,65 @@ def to_one_hot(y, nb_class, dtype=None):
...
+                ". Option 'return_counts=True' works starting"
+                " from version 1.9.0.")
+
+    def make_node(self, x):
+        x = basic.as_tensor_variable(x)
+        #x = x.flatten()
+        
+        outputs = []
+        # output0 = basic.TensorType(broadcastable=[False], dtype=x.dtype)()
+        output0 = x.flatten().type()
```
Comment (nouiz) [inline theano/tensor/extra_ops.py]: It is more clean if you use the line in comment then this one.

File: theano/tensor/extra_ops.py (hunk 5 above)
Comment (nouiz) [inline theano/tensor/extra_ops.py]: You forgot to do this change.

File: theano/tensor/tests/test_extra_ops.py (hunk 6 above)
Comment (nouiz) [inline theano/tensor/tests/test_extra_ops.py]: do this fix.

File: theano/tensor/tests/test_extra_ops.py (hunk 7 above)
Comment (nouiz) [inline theano/tensor/tests/test_extra_ops.py]: and this one.

File: theano/tensor/tests/test_extra_ops.py (hunk 9 above)
Comment (nouiz) [inline theano/tensor/tests/test_extra_ops.py]: this one too need to be done.

**Comment by harlouci** [comment]
Additional corrections on class Unique

File: theano/tensor/extra_ops.py [hunk 12]
```diff
@@ -1006,3 +1005,60 @@ def to_one_hot(y, nb_class, dtype=None):
     ret = theano.tensor.set_subtensor(ret[theano.tensor.arange(y.shape[0]), y],
                                       1)
     return ret
+
+class Unique(theano.Op):
```
Comment (nouiz) [inline theano/tensor/extra_ops.py]: Add a docstring that tell this op isn't implemented on the GPU.

**Comment by harlouci** [comment]
I added a docstring for class Unique

**Comment by nouiz** [comment]
thanks.

File: theano/tensor/extra_ops.py (hunk 12 above)
Comment (nouiz) [inline theano/tensor/extra_ops.py]: Add usage examples here.

**Pull request merged**